### PR TITLE
backport __pairs metamethod to our tweaked version of Lua

### PR DESCRIPTION
**Pull request by ChrisFloofyKitsune**
I copied what I could from the changes to Lua 5.2 to backport making __pairs work as a metamethod.

This greatly improves compatibility with sollua as it assumes that __pairs actually exists and tries to use it
Also makes some of the Lua integrations for RmlUI better to use.

### CODE REVIEW

File: rts/lib/lua/src/lbaselib.cpp [hunk 1]
```diff
@@ -236,10 +236,15 @@ static int luaB_next (lua_State *L) {
 
 
 static int luaB_pairs (lua_State *L) {
-  luaL_checktype(L, 1, LUA_TTABLE);
-  lua_pushvalue(L, lua_upvalueindex(1));  /* return generator, */
-  lua_pushvalue(L, 1);  /* state, */
-  lua_pushnil(L);  /* and initial value */
+  luaL_checkany(L, 1);
+  if (luaL_getmetafield(L, 1, "__pairs") == LUA_TNIL) {  /* no metamethod? */
+    lua_pushcfunction(L, luaB_next);  /* return generator, */
```
Comment (sprunk) [inline rts/lib/lua/src/lbaselib.cpp:241]: Why the change from pushing `lua_upvalueindex(1)` to pushing `luaB_next` directly? Can the generator ever not be `next`? Maybe there are lifetime differences (won't this make a new instance of the C function each time)? What about other functions that use the "push upvalue 1" pattern?

Comment (ChrisFloofyKitsune) [reply rts/lib/lua/src/lbaselib.cpp:241]: the code I have there is (almost) exactly how the latest version of Lua does it.

https://github.com/lua/lua/blob/c1dc08e8e8e22af9902a6341b4a9a9a7811954cc/lbaselib.c#L284-L296

```cpp
static int luaB_pairs (lua_State *L) {
  luaL_checkany(L, 1);
  if (luaL_getmetafield(L, 1, "__pairs") == LUA_TNIL) {  /* no metamethod? */
    lua_pushcfunction(L, luaB_next);  /* will return generator, */
    lua_pushvalue(L, 1);  /* state, */
    lua_pushnil(L);  /* and initial value */
  }
  else {
    lua_pushvalue(L, 1);  /* argument 'self' to metamethod */
    lua_callk(L, 1, 3, 0, pairscont);  /* get 3 values from metamethod */
  }
  return 3;
}
```

Comment (ChrisFloofyKitsune) [reply rts/lib/lua/src/lbaselib.cpp:241]: They've also completely dropped the use of "auxopen" in setting up the base functions.

Which was a convenience method that make a closure so that the other function needed was there was always there as the first upvalue.

```cpp
static void auxopen (lua_State *L, const char *name,
                     lua_CFunction f, lua_CFunction u) {
  lua_pushcfunction(L, u);
  lua_pushcclosure(L, f, 1);
  lua_setfield(L, -2, name);
}
```

I'll also look and see if there was also any changes to how lua_pushcfunction worked in the newer versions of Lua

Comment (ChrisFloofyKitsune) [reply rts/lib/lua/src/lbaselib.cpp:241]: ...
seems that there was a change that added "light C functions", that allowed the use of upvalues to be removed
the idea being that in certain cases C functions could be used without needing to make a full closure

the newest version of pushcclosure is like so- which avoids GC overhead
```cpp
LUA_API void lua_pushcclosure (lua_State *L, lua_CFunction fn, int n) {
  lua_lock(L);
  if (n == 0) {
    setfvalue(s2v(L->top.p), fn);
    api_incr_top(L);
  }
  else {
    int i;
    CClosure *cl;
    api_checkpop(L, n);
    api_check(L, n <= MAXUPVAL, "upvalue index too large");
    cl = luaF_newCclosure(L, n);
    cl->f = fn;
    for (i = 0; i < n; i++) {
      setobj2n(L, &cl->upvalue[i], s2v(L->top.p - n + i));
      /* does not need barrier because closure is white */
      lua_assert(iswhite(cl));
    }
    L->top.p -= n;
    setclCvalue(L, s2v(L->top.p), cl);
    api_incr_top(L);
    luaC_checkGC(L);
  }
  lua_unlock(L);
}
```

this is quite the rabbit hole to potentially go down to backport all this supporting code as well- thanks for bringing up your `won't this make a new instance of the C function each time?` concern.

I'll change the code to use the upvalue since "light C functions" do not seem to be supported in our jank version of Lua

Comment (ChrisFloofyKitsune) [reply rts/lib/lua/src/lbaselib.cpp:241]: It seems that it was realized that c functions... don't need a fully closure in most cases.

**Review by sprunk — APPROVED** [review]
IMO it would be nice to eventually migrate any backend performance changes from future Lua versions that don't affect behaviour.

Perhaps leave a note like
```
/* Meta pairs backported from Lua 5.2, originally 
 * > lua_pushcfunction(L, luaB_next);
 * but requires "light C function" support */
```

Looks good for now though.

File: rts/lib/lua/src/lbaselib.cpp [hunk 2]
```diff
@@ -236,10 +236,15 @@ static int luaB_next (lua_State *L) {
 
 
 static int luaB_pairs (lua_State *L) {
-  luaL_checktype(L, 1, LUA_TTABLE);
-  lua_pushvalue(L, lua_upvalueindex(1));  /* return generator, */
-  lua_pushvalue(L, 1);  /* state, */
-  lua_pushnil(L);  /* and initial value */
+  luaL_checkany(L, 1);
+  if (luaL_getmetafield(L, 1, "__pairs") == LUA_TNIL) {  /* no metamethod? */
+    lua_pushvalue(L, lua_upvalueindex(1));  /* return generator, */
```
Comment (sprunk) [inline rts/lib/lua/src/lbaselib.cpp:241]: Perhaps this could have `luaL_checktype(L, 1, LUA_TTABLE);`? I'd expect that to produce a better error message ("error in pairs...") than letting it fail inside `next` (since the next is there implicitly).

Comment (ChrisFloofyKitsune) [reply rts/lib/lua/src/lbaselib.cpp:248]: I think it failing inside `next` will be fine- it having failed in `pairs` will just be one level up the callstack

I can't move the `luaL_checktype(L, 1, LUA_TTABLE);` from `next` since that is sometimes used on it's own in Lua code, and checking the type twice in a row seems... silly (and also inefficient)

**Comment by lhog** [comment]
Looks good. Makes our Lua even weirder than it is currently, but given the very likely possibility we stuck with 5.1 Lua forever, maybe it's not that bad.

**Comment by ChrisFloofyKitsune** [comment]
> Looks good. Makes our Lua even weirder than it is currently, but given the very likely possibility we stuck with 5.1 Lua forever, maybe it's not that bad.

I if heard right, we're stuck with what we have because stuff was changed for the sake of making sure gadgets stay synced.

Bring our Lua implementation and the existing Lua code up to date is... unlikely to happen as it's a major project to update it while keeping the modifications made to have it work for what we need.

**Comment by lhog** [comment]
> I if heard right, we're stuck with what we have because stuff was changed for the sake of making sure gadgets stay synced.
> 
> Bring our Lua implementation and the existing Lua code up to date is... unlikely to happen as it's a major project to update it while keeping the modifications made to have it work for what we need.

It's a combination of two factors:
1) Custom changes to the vanilla Lua with no sane tracking where changes occurred. Yes some of these are easy to port, others not so much
2) Amount of changes along the 5.1 --> 5.4 route.

I'm not good with Lua, so after a day worth of efforts and struggles I gave up. Doesn't mean it can't be done by someone more qualified.
